iroha/src/sumeragi.rs

```diff
@@ -698,52 +740,38 @@ impl InitializedNetworkTopology {
         &self,
         signature: Signature,
         role: Role,
-        message: &[u8],
+        message_payload: &[u8],
     ) -> Result<(), String> {
-        match role {
-            Role::ProxyTail => {
-                if signature.public_key == self.proxy_tail().public_key {
-                    Ok(())
-                } else {
-                    Err("Proxy tail has a different public key.".to_string())
-                }
-            }
-            Role::Leader => {
-                if signature.public_key == self.leader().public_key {
-                    Ok(())
-                } else {
-                    Err("Leader has a different public key.".to_string())
-                }
-            }
-            Role::ValidatingPeer => {
-                if self
-                    .validating_peers()
-                    .iter()
-                    .any(|peer| peer.public_key == signature.public_key)
-                {
-                    Ok(())
-                } else {
-                    Err("No validating peer with this public key exists.".to_string())
-                }
-            }
-            Role::ObservingPeer => {
-                if self
-                    .peers_set_b()
-                    .iter()
-                    .any(|peer| peer.public_key == signature.public_key)
-                {
-                    Ok(())
-                } else {
-                    Err("No observing peer with this public key exists.".to_string())
-                }
-            }
+        if role
+            .peers(&self)
+            .iter()
+            .any(|peer| peer.public_key == signature.public_key)
+        {
+            Ok(())
+        } else {
+            Err(format!("No {:?} with this public key exists.", role))
         }
-        .and(signature.verify(message))
+        .and(signature.verify(message_payload))
+    }
+
+    /// Returns signatures of the peers with the specified `roles` from all `signatures`.
+    pub fn signatures_for_roles(&self, roles: &[Role], signatures: &[Signature]) -> Vec<Signature> {
```

Comment:
1. If we have `Signatures`, maybe it's better to use it instead of `Vec<Signature>`
2. Naming is not good, we filter `signatures` by `roles` entities so we should have `filter` in the method name.

Reply:
I will check if it makes sense to use `Signatures` here. With naming, yes, I will rename it.